feat: add steps to deploying to the RKE2 custom cluster (#525)
harvester/docs: docs/rancher/cloud-provider.md

1. Select `Harvester(Out-of-tree)` option.

    ![](/img/v1.2/rancher/rke-cloud-provider.png)

2. Install `Harvester Cloud Provider` from the Rancher marketplace.

    ![](/img/v1.2/rancher/install-harvester-cloud-provider.png)

### Deploying to the RKE2 Cluster with Harvester Node Driver

When spinning up an RKE2 cluster using the Harvester node driver, select the `Harvester` cloud provider. The node driver will then help deploy both the CSI driver and CCM automatically.

  ![](/img/v1.2/rancher/rke2-cloud-provider.png)

### Deploying to the Custom RKE2 Cluster with Harvester Cloud Provider (Experimental)

![](/img/v1.2/rancher/custom.png)
1. Generate cloud config data using the script `generate_addon.sh`, and then place the data on every custom node (directory: `/etc/kubernetes/cloud-config`).

    ```bash
    curl -sfL https://raw.githubusercontent.com/harvester/cloud-provider-harvester/master/deploy/generate_addon.sh | bash -s <serviceaccount name> <namespace>
    ```

    :::note

>     The `generate_addon.sh` script depends on `kubectl` and `jq` to operate the Harvester cluster.

Reviewer: ```suggestion
    The script depends on `kubectl` and `jq` when operating the Harvester cluster, and functions only when given access to the `Harvester Cluster` kubeconfig file.
```

>     The script needs access to the `Harvester Cluster` kubeconfig to work. You can find the `kubeconfig` file from one of the Harvester management nodes in the `/etc/rancher/rke2/rke2.yaml` path. The content is like following. Remember to change server IP to the VIP address.

Reviewer: ```suggestion
    You can find the `kubeconfig` file in one of the Harvester management nodes in the `/etc/rancher/rke2/rke2.yaml` path. The server IP must be replaced with the VIP address.

  Example of content:
```

    ```yaml
    apiVersion: v1
    clusters:
    - cluster:
        certificate-authority-data: <redacted>
        server: https://127.0.0.1:6443
      name: default
    # ...
    ```

>     The namespace needs to be the namespace in which the guest cluster will be created.

Reviewer: ```suggestion
    You must specify the namespace in which the guest cluster will be created.
```

    :::

>     The output will look as follows:

Reviewer: ```suggestion
    Example of output:
```

    ```
    curl -sfL https://raw.githubusercontent.com/harvester/cloud-provider-harvester/master/deploy/generate_addon.sh | bash -s <serviceaccount name> <namespace>
    ```

    The output will look as follows:

    ```
    ########## cloud config ############
    apiVersion: v1
    clusters:
    - cluster:
        certificate-authority-data: <CACERT>
        server: https://HARVESTER-ENDPOINT/k8s/clusters/local
      name: local
    contexts:
    - context:
        cluster: local
        namespace: default
        user: harvester-cloud-provider-default-local
      name: harvester-cloud-provider-default-local
    current-context: harvester-cloud-provider-default-local
    kind: Config
    preferences: {}
    users:
    - name: harvester-cloud-provider-default-local
      user:
        token: <TOKEN>

    ########## cloud-init user data ############
    write_files:
    - encoding: b64
      content: <CONTENT>
      owner: root:root
      path: /etc/kubernetes/cloud-config
      permissions: '0644'
    ```

    :::note

    The `generate_addon.sh` script depends on `kubectl` and `jq` to operate the Harvester cluster.

>     The script needs access to the `Harvester Cluster` kubeconfig to work. You can find the `kubeconfig` file from one of the Harvester management nodes in the `/etc/rancher/rke2/rke2.yaml` path.

Reviewer 2: We could add a kindly reminder that the kubeconfig file `/etc/rancher/rke2/rke2.yaml` on the management node looks like the following:

```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <redacted>
    server: https://127.0.0.1:6443
  name: default
```

It's error-prone that if the user executes the `generate_addon.sh` on the node directly with the kubeconfig file's `server` field unmodified. The generated cloud-config contains `127.0.0.1` which is not going to work later on. It needs to be the Harvester cluster's VIP address.

Author: I added more content to this block. Please take a look again. Thank you.

1. Select `Harvester(Out-of-tree)` option.

    ![](/img/v1.2/rancher/rke-cloud-provider.png)

2. Install `Harvester Cloud Provider` from the Rancher marketplace.

    ![](/img/v1.2/rancher/install-harvester-cloud-provider.png)

### Deploying to the RKE2 Cluster with Harvester Node Driver

When spinning up an RKE2 cluster using the Harvester node driver, select the `Harvester` cloud provider. The node driver will then help deploy both the CSI driver and CCM automatically.

  ![](/img/v1.2/rancher/rke2-cloud-provider.png)

> ### Deploying to the RKE2 custom cluster with Harvester Cloud Provider (experimental)

Reviewer: ```suggestion
### Deploying to the Custom RKE2 Cluster with Harvester Cloud Provider (Experimental)
```

![](/img/v1.2/rancher/custom.png)

> 1. Use `generate_addon.sh` to generate a cloud-config and place it into the directory `/etc/kubernetes/cloud-config` on every custom node.

Reviewer: ```suggestion
1. Generate cloud config data using the script `generate_addon.sh`, and then place the data on every custom node (directory: `/etc/kubernetes/cloud-config`).
```

    ```
    curl -sfL https://raw.githubusercontent.com/harvester/cloud-provider-harvester/master/deploy/generate_addon.sh | bash -s <serviceaccount name> <namespace>
    ```

>     The output will look as follows:

Reviewer: ```suggestion
  :::note
  The script depends on `kubectl` and `jq` when operating the Harvester cluster, and functions only when given access to the `Harvester Cluster` kubeconfig file. You can find the kubeconfig file in one of the Harvester management nodes (path: `/etc/rancher/rke2/rke2.yaml`).

  You must specify the namespace in which the guest cluster will be created.
  :::

  Example of output:
```

Reviewer: I moved the note here because the information must be closer to the main instruction and the actual command.

Author: Hi @jillian-maroket, I move the note to close to the command. Please take a look again. Thank you.

    ```
    ########## cloud config ############
    apiVersion: v1
    clusters:
    - cluster:
        certificate-authority-data: <CACERT>
        server: https://HARVESTER-ENDPOINT/k8s/clusters/local
      name: local
    contexts:
    - context:
        cluster: local
        namespace: default
        user: harvester-cloud-provider-default-local
      name: harvester-cloud-provider-default-local
    current-context: harvester-cloud-provider-default-local
    kind: Config
    preferences: {}
    users:
    - name: harvester-cloud-provider-default-local
      user:
        token: <TOKEN>

    ########## cloud-init user data ############
    write_files:
    - encoding: b64
      content: <CONTENT>
      owner: root:root
      path: /etc/kubernetes/cloud-config
      permissions: '0644'
    ```

>     :::note

Reviewer: Please remove this. In another comment, I suggested placing the note after the command.

    The `generate_addon.sh` script depends on `kubectl` and `jq` to operate the Harvester cluster.

    The script needs access to the `Harvester Cluster` kubeconfig to work. You can find the `kubeconfig` file from one of the Harvester management nodes in the `/etc/rancher/rke2/rke2.yaml` path.

    The namespace needs to be the namespace in which the guest cluster will be created.

    :::

> 2. Create a VM in the Harvester cluster.

Reviewer: ```suggestion
1. Create a VM in the Harvester cluster with the following settings:
```

>   2-1. The VM minimum requirements are 2 vCPUs and 4GB of RAM. Disk space depends on different VMImage.
> 
>     ![](/img/v1.3/rancher/custom-cluster-vm-cpu-and-ram.png)
> 
>   2-2. The Network name should be `nic-<number>`.
> 
>     ![](/img/v1.3/rancher/custom-cluster-vm-network.png)
> 
>   2-3. Copy and paste the `cloud-init user data` content to Advanced Options.
> 
>     ![](/img/v1.3/rancher/custom-cluster-vm-user-data.png)
> 

Reviewer: ```suggestion
  - **Basics** tab: The minimum requirements are 2 CPUs and 4 GiB of RAM. The required disk space depends on the VM image.

    ![](/img/v1.3/rancher/custom-cluster-vm-cpu-and-ram.png)

  - **Networks** tab: Specify a network name with the format `nic-<number>`.

    ![](/img/v1.3/rancher/custom-cluster-vm-network.png)

  - **Advanced Options** tab: Copy and paste the content of the **Cloud Config User Data** screen.

    ![](/img/v1.3/rancher/custom-cluster-vm-user-data.png)
```

> 3. Configure the **Cloud Provider** to `Harvester` and select **Create** to spin up the cluster.
>      ![](/img/v1.2/rancher/create-custom-rke2.png)

Reviewer: ```suggestion
1. On the **Basics** tab of the **Cluster Configuration** screen, select **Harvester** as the **Cloud Provider** and then select **Create** to spin up the cluster.

  ![](/img/v1.2/rancher/create-custom-rke2.png)
```

> 4. Follow the steps in Registration tab to run RKE2 installation command on the VM.
> 
>     ![](/img/v1.3/rancher/custom-cluster-registration.png)

Reviewer: ```suggestion
1. On the **Registration** tab, perform the steps required to run the RKE2 registration command on the VM.

    ![](/img/v1.3/rancher/custom-cluster-registration.png)
```

      name: local
    contexts:
    - context:
        cluster: local
        namespace: default
        user: harvester-cloud-provider-default-local
      name: harvester-cloud-provider-default-local
    current-context: harvester-cloud-provider-default-local
    kind: Config
    preferences: {}
    users:
    - name: harvester-cloud-provider-default-local
      user:
        token: <TOKEN>

    ########## cloud-init user data ############
    write_files:
    - encoding: b64
      content: <CONTENT>
      owner: root:root
      path: /etc/kubernetes/cloud-config
      permissions: '0644'
    ```

1. Create a VM in the Harvester cluster with the following settings:

  - **Basics** tab: The minimum requirements are 2 CPUs and 4 GiB of RAM. The required disk space depends on the VM image.

    ![](/img/v1.3/rancher/custom-cluster-vm-cpu-and-ram.png)

  - **Networks** tab: Specify a network name with the format `nic-<number>`.

    ![](/img/v1.3/rancher/custom-cluster-vm-network.png)

  - **Advanced Options** tab: Copy and paste the content of the **Cloud Config User Data** screen.

    ![](/img/v1.3/rancher/custom-cluster-vm-user-data.png)

> 1. On the **Basics** tab of the **Cluster Configuration** screen, select **Harvester** as the **Cloud Provider** and then select **Create** to spin up the cluster.

Reviewer 3: In the [doc preview](https://66863d5f42fbf17e30f4e778--harvester-preview.netlify.app/v1.4/rancher/cloud-provider#deploying-to-the-rke2-custom-cluster-experimental), the ordered list starts with `1.` again. I'm not sure if the intention is to continue with `3.`. And we should mention here the user should go to the Rancher's GUI, but not Harvester's GUI anymore.

Please also add the doc source to v1.3/v1.2 versions if needed (it's currently only in the dev version).

Author: Fixed it and also migrated the document to v1.3/v1.2. Thanks.
